### PR TITLE
[#599] Replace Apache with nginx and run services as non-root, with t…

### PR DESCRIPTION
**Pull request by justafish**
…he new users having the same UID/GID as the host user to ensure correct volume permissions

## Issue

https://github.com/jsdrupal/drupal-admin-ui/issues/599

## Testing instructions

- Remove any untracked files (`git clean -fdX`)
- Rebuild the docker containers with the new command mentioned in the README
- Bring up the containers, enter them and check mounted volumes and their files are using the non-root user
- Outside of the containers, check the files created in mounted volumes are owned by your host user

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ Move it to ~/.cache/yarn (see https://yarnpkg.com/lang/en/docs/cli/cache/ for in
 ```
 git clone git@github.com:jsdrupal/drupal-admin-ui.git
 cd drupal-admin-ui
+uid=$(id -u) gid=$(id -g) docker-compose buil
```
Comment (dawehner) [inline README.md]:
```suggestion
uid=$(id -u) gid=$(id -g) docker-compose build
```

What happens with windows machines here? I could imagine `id` is not available there?

Comment (justafish) [reply README.md]: @dawehner I'm not sure :thinking:

**Comment by dawehner** [comment]
Note: When trying to build the container for this PR I run into the following error message:

```
groupadd: GID '100' already exists
ERROR: Service 'node' failed to build: The command '/bin/sh -c mkdir -p /var/www &&     deluser node &&     sed -i 's/^CREATE_MAIL_SPOOL=yes/CREATE_MAIL_SPOOL=no/' /etc/default/useradd &&     groupadd -g ${NODE_GID} node &&     useradd -u ${NODE_UID} -d /var/www -g ${NODE_GID} node &&     sed -e 's/# %wheel ALL=(ALL) NOPASSWD: ALL/%wheel ALL=(ALL) NOPASSWD: ALL/g' -i /etc/sudoers &&     usermod -a -G wheel node &&     chown node:node /var/www' returned a non-zero code: 4
```

GID is the group of my current user. Here is the output of the `id` command to see all the users / groups:
```
$ id
uid=1000(dawehner) gid=100(users) groups=100(users),1(wheel),17(audio),57(networkmanager),131(docker)
```

**Comment by martinfrances107** [comment]
I was unable to reproduce dawehner's error on this branch.

I got it to work ONLY after the following 'nuke it from orbit' script

`docker-compose down` - // removes containers and network
`docker-compose ps -a `// parnanoid check that containers are gone
`docker  volume prune` // wipe away any stale volume 
`docker volume list` // paranoid check that the volumes are wiped.

and the following the updated README.md command for building 

`uid=$(id -u) gid=$(id -g) docker-compose build`

Anyway I hope this helps ... the new fangled command is all well and good - but to my mind it is too complicated.. If I were manning a support desk then I would expect to get lots of calls about this command.

I am failure with the wodby project -- which provides open source drupal containers   
https://github.com/wodby/drupal

They have abstracted there core commands into a Makefile so the user types 

make build
make logs
make up

etc and all the complexities involving settings permissions can be automated.

https://github.com/wodby/drupal/blob/master/8/Makefile

This PR has lots of complexities so I want to streamline the review process by saying I will work on the Makefile stuff as a followup issue.

**Comment by martinfrances107** [comment]
After playing around with this branch some more.

when I visit http://127.0.0.1

I can see a little screw up concerning png and jpg files

in chrome's devtools -- console window I see 9 errors with http code 500 "Internal server error"

below I am quoting just 2 of the errored links.

within .docker/drupal/nginx.default.conf I can see setup for

image files

```
location ~* \.(js|css|png|jpg|jpeg|gif|ico|svg)$ {
    try_files $uri @rewrite;
    expires max;
    log_not_found off;
}
```
but I think there is a conflict with

```
location ~ ^/sites/.*/files/styles/ {
    try_files $uri @rewrite;
}
```
Anyway I will try and fix this tomorrow.

```
http://127.0.0.1/sites/default/files/styles/medium_8_7/public/umami-bundle.png?itok=92iAaAEI

http://127.0.0.1/sites/default/files/styles/scale_crop_7_3_large/public/veggie-pasta-bake-hero-umami.jpg?itok=Z-0248WD
```

**Comment by martinfrances107** [comment]
In the comment above I point out a specific "location" condition that was causing problems

I put our nginx.conf file into this website https://nginx.viraptor.info/
which can be used to test routes.... this confirmed my visual analysis of the problem.

I don't know the source of the nginx.conf file but it looked failure, so I googled it and found this.
https://www.nginx.com/resources/wiki/start/topics/recipes/drupal/

My only comment is that the umami demo install does not work with this sample nginx template.
So I am going to try and fix this upstream. umami is a experimental D8 core module.
So core is the place to dicuss this and place a  nginx.conf.dist file -- maybe?

In core this issue will get more eyes.. and a a better solution. All the solutions I can think of are a bit hacky.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+build:
+	uid=$$(id -u) gid=$$(id -g) docker-compose build
+
+up:
+	docker-compose up
+
+clean:
+	docker-compose rm -vsf
+	docker-compose down -v --remove-orphans
+	sudo rm -f demo/docroot/sites/default/settings.php
```
Comment (martinfrances107) [inline Makefile]: Thank you this Makefile is a great improvement.

May I suggest a minor tweak from field testing ... I had a broken build in which a yarn-error.log was generated.
My expectation is that a "make clean" should clean this up... otherwise if the next developer goes through a cycle of make build .. see error 1... fix error 1 ... make build .. move onto error2  see the stale yarn-error.log from error1 and get misdirected about the nature of error2.

It is a bit contrived hey I have already had that one bite me in the behind.  

so in short 

rm yarn-error.log

**Comment by martinfrances107** [comment]
So I have had another mini review after justafish's raft of adjustments.

I can confirm it fixes the problems I reported above :tada: 

Broken image link when visiting the main landing page ( 127.0.0.1/ ). 

But in the new scheme of things I find the node container exits soon after creation 

with 

drupal_admin_ui_node | error An unexpected error occurred: "EACCES: permission denied, mkdir '/var/www/drupal-admin-ui/node_modules/@drupal'".

So the permission setup is not yet bullet proof.

Below I am just posting the step to reproduce and the logs of the node container.

After

> make clean 
> make up
> docker-compose  logs node

WARNING: The uid variable is not set. Defaulting to a blank string.
WARNING: The gid variable is not set. Defaulting to a blank string.
Attaching to drupal_admin_ui_node
drupal_admin_ui_node | yarn install v1.12.3
drupal_admin_ui_node | [1/5] Validating package.json...
drupal_admin_ui_node | [2/5] Resolving packages...
drupal_admin_ui_node | [3/5] Fetching packages...
drupal_admin_ui_node | info fsevents@1.2.4: The platform "linux" is incompatible with this module.
drupal_admin_ui_node | info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
drupal_admin_ui_node | [4/5] Linking dependencies...
drupal_admin_ui_node | warning "workspace-aggregator-54c2a48e-1902-4b93-8b10-07e8340a83ac > admin-ui > react-emotion@9.2.12" has incorrect peer dependency "emotion@^9.1.3".
drupal_admin_ui_node | warning "workspace-aggregator-54c2a48e-1902-4b93-8b10-07e8340a83ac > admin-ui > @storybook/react@4.0.4" has unmet peer dependency "babel-loader@^7.0.0 || ^8.0.0".
drupal_admin_ui_node | warning "workspace-aggregator-54c2a48e-1902-4b93-8b10-07e8340a83ac > admin-ui > @storybook/react > @babel/preset-flow@7.0.0" has unmet peer dependency "@babel/core@^7.0.0-0".
drupal_admin_ui_node | warning "workspace-aggregator-54c2a48e-1902-4b93-8b10-07e8340a83ac > admin-ui > @storybook/react > @storybook/core@4.0.4" has unmet peer dependency "babel-loader@^7.0.0 || ^8.0.0".
drupal_admin_ui_node | warning "workspace-aggregator-54c2a48e-1902-4b93-8b10-07e8340a83ac > admin-ui > @storybook/react > @babel/preset-flow > @babel/plugin-transform-flow-strip-types@7.0.0" has unmet peer dependency "@babel/core@^7.0.0-0".
drupal_admin_ui_node | error An unexpected error occurred: "EACCES: permission denied, mkdir '/var/www/drupal-admin-ui/node_modules/@drupal'".
drupal_admin_ui_node | info If you think this is a bug, please open a bug report with the information provided in "/var/www/drupal-admin-ui/yarn-error.log".
drupal_admin_ui_node | info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
drupal_admin_ui_node | yarn workspace v1.12.3
drupal_admin_ui_node | yarn run v1.12.3
drupal_admin_ui_node | $ react-scripts build
drupal_admin_ui_node | /bin/sh: react-scripts: not found
drupal_admin_ui_node | error Command failed with exit code 127.
drupal_admin_ui_node | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
drupal_admin_ui_node | error Command failed.
drupal_admin_ui_node | Exit code: 1
drupal_admin_ui_node | Command: /usr/local/bin/node
drupal_admin_ui_node | Arguments: /opt/yarn-v1.12.3/lib/cli.js build
drupal_admin_ui_node | Directory: /var/www/drupal-admin-ui/packages/admin-ui
drupal_admin_ui_node | Output:
drupal_admin_ui_node | 
drupal_admin_ui_node | info Visit https://yarnpkg.com/en/docs/cli/workspace for documentation about this command.
drupal_admin_ui_node | yarn workspace v1.12.3
drupal_admin_ui_node | yarn run v1.12.3
drupal_admin_ui_node | $ react-scripts start
drupal_admin_ui_node | /bin/sh: react-scripts: not found
drupal_admin_ui_node | error Command failed with exit code 127.
drupal_admin_ui_node | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
drupal_admin_ui_node | error Command failed.
drupal_admin_ui_node | Exit code: 1
drupal_admin_ui_node | Command: /usr/local/bin/node
drupal_admin_ui_node | Arguments: /opt/yarn-v1.12.3/lib/cli.js start
drupal_admin_ui_node | Directory: /var/www/drupal-admin-ui/packages/admin-ui
drupal_admin_ui_node | Output:
drupal_admin_ui_node | 
drupal_admin_ui_node | info Visit https://yarnpkg.com/en/docs/cli/workspace for documentation about this command.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,49 @@
+SHELL := /bin/sh
```
Comment (dawehner) [inline Makefile]: Let's re-iterate on that, given it won't work on windows.Symfony 4 started using makefiles but then realized it is too much of a huzzle on windows.

**Comment by justafish** [comment]
I'll do some work in a follow-up PR to try and improve this for Mac users. I'll look at Windows after that.

File: .docker/drupal/supervisord.conf [hunk 4]
```diff
@@ -0,0 +1,19 @@
+[supervisord]
+nodaemon=true
+pidfile=/run/supervisord.pid
```
[the remark this answers is not in the document]
Comment (dawehner) [reply .docker/drupal/supervisord.conf]: I was starting to think about: What is alpine using instead of systemD, and well, it is using openRC, which is a tool originiated in gentoo land.

**Comment by martinfrances107** [comment]
I fixed my permissions problems by having a through clean of 

rm -rf  node_modules packages 

I was running on localhost and then as docker and I think this was screwing up the permissions by leaving some files with a root user.

The only true "make clean", maybe to wipe everything and do a git clone again

That is not to say everything is not sweetness and light - the node container still exits with error codes.

here is the current output from 

docker-compose logs node 

drupal_admin_ui_node | yarn run v1.12.3
drupal_admin_ui_node | $ react-scripts start
drupal_admin_ui_node | Starting the development server...
drupal_admin_ui_node | 
drupal_admin_ui_node | events.js:167
drupal_admin_ui_node |       throw er; // Unhandled 'error' event
drupal_admin_ui_node |       ^
drupal_admin_ui_node | 
drupal_admin_ui_node | Error: ENOSPC: System limit for number of file watchers reached, watch '/var/www/drupal-admin-ui/packages/admin-ui/public'
drupal_admin_ui_node |     at FSWatcher.start (internal/fs/watchers.js:165:26)
drupal_admin_ui_node |     at Object.watch (fs.js:1254:11)
drupal_admin_ui_node |     at createFsWatchInstance (/var/www/drupal-admin-ui/node_modules/chokidar/lib/nodefs-handler.js:37:15)
drupal_admin_ui_node |     at setFsWatchListener (/var/www/drupal-admin-ui/node_modules/chokidar/lib/nodefs-handler.js:80:15)
drupal_admin_ui_node |     at FSWatcher.NodeFsHandler._watchWithNodeFs (/var/www/drupal-admin-ui/node_modules/chokidar/lib/nodefs-handler.js:232:14)
drupal_admin_ui_node |     at FSWatcher.NodeFsHandler._handleDir (/var/www/drupal-admin-ui/node_modules/chokidar/lib/nodefs-handler.js:414:19)
drupal_admin_ui_node |     at FSWatcher.<anonymous> (/var/www/drupal-admin-ui/node_modules/chokidar/lib/nodefs-handler.js:462:19)
drupal_admin_ui_node |     at FSWatcher.<anonymous> (/var/www/drupal-admin-ui/node_modules/chokidar/lib/nodefs-handler.js:467:16)
drupal_admin_ui_node |     at FSReqWrap.oncomplete (fs.js:155:5)
drupal_admin_ui_node | Emitted 'error' event at:
drupal_admin_ui_node |     at FSWatcher._handleError (/var/www/drupal-admin-ui/node_modules/chokidar/index.js:260:10)
drupal_admin_ui_node |     at createFsWatchInstance (/var/www/drupal-admin-ui/node_modules/chokidar/lib/nodefs-handler.js:39:5)
drupal_admin_ui_node |     at setFsWatchListener (/var/www/drupal-admin-ui/node_modules/chokidar/lib/nodefs-handler.js:80:15)
drupal_admin_ui_node |     [... lines matching original stack trace ...]
drupal_admin_ui_node |     at FSReqWrap.oncomplete (fs.js:155:5)
drupal_admin_ui_node | error Command failed with exit code 1.
drupal_admin_ui_node | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
drupal_admin_ui_node | error Command failed.
drupal_admin_ui_node | Exit code: 1
drupal_admin_ui_node | Command: /usr/local/bin/node
drupal_admin_ui_node | Arguments: /opt/yarn-v1.12.3/lib/cli.js start
drupal_admin_ui_node | Directory: /var/www/drupal-admin-ui/packages/admin-ui
drupal_admin_ui_node | Output:
drupal_admin_ui_node | 
drupal_admin_ui_node | info Visit https://yarnpkg.com/en/docs/cli/workspace for documentation about this command.

**Comment by martinfrances107** [comment]
PS I think this line in the  .docker/node/Dockerfile is harmful.

`mkdir -p /var/www/drupal-admin-ui `

it is both 

A) In the dockerfile 

and 

B) a mount point from the volume line in the docker-compose.yml file

stack-overflow has lots of stories of errors where mkdir clashes with dynamic mount points.

Here is the mkdir in context.
 

```
USER node

#RUN mkdir -p /var/www/drupal-admin-ui
WORKDIR /var/www/drupal-admin-ui
```

**Comment by martinfrances107** [comment]
Returning to the error

`drupal_admin_ui_node | Error: ENOSPC: System limit for number of file watchers reached, watch '/var/www/drupal-admin-ui/packages/admin-ui/public'
`
its is a inode exhaustion issue

http://www.jagregory.com/writings/docker-container-out-of-space/

When I put the **find** command from the blog into the node setup.sh file like this 
I get the following inode report 

```
sudo chown -R node:node /var/www/.cache
yarn install
find . -xdev -type f | cut -d "/" -f 2 | sort | uniq -c | sort -n
yarn workspace admin-ui build

# Create a file to indicate build has finished
touch /var/www/.yarn-build-complete

yarn workspace admin-ui start
```
```
drupal_admin_ui_node |       1 .editorconfig
drupal_admin_ui_node |       1 .eslintrc.json
drupal_admin_ui_node |       1 .github
drupal_admin_ui_node |       1 .gitignore
drupal_admin_ui_node |       1 .prettierrc.json
drupal_admin_ui_node |       1 CODE_OF_CONDUCT.md
drupal_admin_ui_node |       1 CRA-README.md
drupal_admin_ui_node |       1 LICENSE.md
drupal_admin_ui_node |       1 Makefile
drupal_admin_ui_node |       1 README.md
drupal_admin_ui_node |       1 VALUES.md
drupal_admin_ui_node |       1 docker-compose.yml
drupal_admin_ui_node |       1 greenkeeper.json
drupal_admin_ui_node |       1 package.json
drupal_admin_ui_node |       1 yarn.lock
drupal_admin_ui_node |       2 .circleci
drupal_admin_ui_node |       4 nbproject
drupal_admin_ui_node |       8 .docker
drupal_admin_ui_node |      57 admin_ui_support
drupal_admin_ui_node |      83 .git
drupal_admin_ui_node |    4943 packages
drupal_admin_ui_node |   23799 demo
drupal_admin_ui_node |   55522 node_modules
```

**Comment by martinfrances107** [comment]
I managed resolve the inode issue.

Although I do not have the final solution. The blogs post I talked about  essentially says "find the tmp directory clean it up and your done." It looked like we had to clean up packages, demo or node_modules

my solution was to clear the cache before changing the ownership of .cache dir in the setup.sh file

```
if [ ! -f /var/www/drupal-admin-ui/packages/admin-ui/.env.local ]; then
    cp packages/admin-ui/.env packages/admin-ui/.env.local
    chmod 666 packages/admin-ui/.env.local
fi

yarn clean:cache

sudo chown -R node:node /var/www/.cache
yarn install
```
Now my node container stays up ... and the application springs into life.

I can see other errors hiding behind this one related to the toolbar ( on localhost:3000/ )  .. but I will try and provide a solution first rather than flooding this PR with field reports.

**Comment by justafish** [comment]
I've removed the sharing of the user's local composer/yarn cache and created a new one as a container volume. I've also added some more thorough cleaning to the `make clean` command.

**Comment by mattgrill** [comment]
One thing I wanted to point out was that the image sizes for `drupal-admin-ui_node` and `drupal-admin-ui_drupal` are at 24.4GB. That seems quite large.

**Comment by justafish** [comment]
@mattgrill thanks for the review! The images are large because each step building the image creates a new layer, and since it's a local image you have all of them.

What happens if you directly run the command that `make build` is trying to run?

IIRC the built-in docker packages for Ubuntu aren't very good. Try this: https://gist.github.com/justafish/5679be8cb6ccc37bf1f266ffa2f5c4bf#docker

**Comment by martinfrances107** [comment]
Ubuntu 18.04.1 LTS

```
docker images 
REPOSITORY               TAG                 IMAGE ID            CREATED             SIZE
drupal-admin-ui_drupal   latest              11341395baef        12 hours ago        128MB
drupal-admin-ui_node     latest              4f00af285c40        12 hours ago        349MB
```

A) As a minor thing that can be resolved in a followup issue.
drupal-admin-ui/demo/docroot/.htaccess sets the current policy about when to gzip.
but the nginx.conf does not contain any reference to gzipping 

B) I think the 

RUN mkdir -p /var/www/drupal-admin-ui

is redundant but from clean install it works with and without.

As a mini review are issue are resolved -- all the issues I see when I look hard can be seen on the master branch.

 So for it is worth this PR looks good.... sorry if my constant bugs and updated have seemed a bit much.

**Comment by martinfrances107** [comment]
Wow that was quick

A) Visually all the changes look good. in both the nginx.conf and the dev tools network tab.
B) Chrome's lighthouse drops the appropriate warnings 
C) Through automated checking there are no integrity warnings the nginx.conf  as tested by : -

make drupal:cli
sudo nginx -t  

in short thank you...

**Comment by mattgrill** [comment]
This works for me. _NICE_

File: .docker/drupal/setup.sh [hunk 5]
```diff
@@ -1,31 +1,47 @@
...
+    echo "\$config_directories['sync'] = '../config/sync';" >> docroot/sites/default/settings.php
     mkdir docroot/sites/default/files docroot/sites/simpletest reports
-    chown apache:apache docroot/sites/default/files docroot/sites/simpletest reports
     chmod 777 docroot/sites/default/files
-    composer config repositories.repo-name path "/var/www/admin_ui_support"
-    composer require justafish/drupal-admin-ui-support:dev-master
     composer install
-    chmod +x /var/www/drupal/docroot/core/scripts/drupal
+    composer config repositories.repo-name path "/var/www/admin_ui_support"
+    COMPOSER_MEMORY_LIMIT=-1 composer require justafish/drupal-admin-ui-support:dev-master
```
Comment (dawehner) [inline .docker/drupal/setup.sh]: I'm wondering, can/should we use cross-env here?

File: .docker/drupal/setup.sh [hunk 6]
```diff
@@ -34,6 +50,5 @@ echo "##########################################################################
 ##############################################################################################"
 
 # Create a file to indicate installation has finished
-touch /root/.drupal-installation-complete
-
-apachectl start -DFOREGROUND
+touch ~/.drupal-installation-complete
```
Comment (dawehner) [inline .docker/drupal/setup.sh]: I'm wondering, can we use the pidfile from the upservisord config for this instead?

File: .docker/drupal/setup.sh (hunk 5 above)
Comment (justafish) [reply .docker/drupal/setup.sh]: This only runs in the container, so it's not necessary

File: .docker/drupal/setup.sh (hunk 6 above)
Comment (justafish) [reply .docker/drupal/setup.sh]: happy to look at that in a follow up

File: .docker/drupal/setup.sh (hunk 5 above)
Comment (dawehner) [reply .docker/drupal/setup.sh]: good point. Thank you

File: .docker/drupal/setup.sh (hunk 6 above)
Comment (dawehner) [reply .docker/drupal/setup.sh]: Sure! Create a "easy-pickup" task :)
